core/src/main.cpp

```diff
@@ -547,15 +548,25 @@ auto main(int _argc, char* _argv[]) -> int
 		log::trace("Initializing listening socket (host=[{}], port=[{}]).", address.to_string(), port);
 		std::make_shared<listener>(ioc, tcp::endpoint{address, port}, config)->run();
 
-		// Capture SIGINT and SIGTERM to perform a clean shutdown.
+		// SIGINT and SIGTERM instruct the server to shut down.
+		// Ignore SIGPIPE. The iRODS networking code assumes SIGPIPE is ignored or caught.
 		log::trace("Initializing signal handlers.");
-		net::signal_set signals{ioc, SIGINT, SIGTERM};
-		signals.async_wait([&ioc](const beast::error_code&, int _signal) {
-			// Stop the io_context. This will cause run() to return immediately, eventually destroying the
-			// io_context and all of the sockets in it.
-			log::warn("Received signal [{}]. Shutting down.", _signal);
-			ioc.stop();
-		});
+		net::signal_set signals{ioc, SIGINT, SIGTERM, SIGPIPE};
+
+		const std::function<void(const beast::error_code&, int)> process_signals =
+			[&ioc, &signals, &process_signals](const beast::error_code&, int _signal) {
+				if (SIGPIPE == _signal) {
+					signals.async_wait(process_signals);
+					return;
+				}
+
+				// Stop the io_context. This will cause run() to return immediately, eventually destroying
+				// the io_context and all of the sockets in it.
+				log::warn("Received signal [{}]. Shutting down.", _signal);
+				ioc.stop();
+			};
+
+		signals.async_wait(process_signals);
```

Comment:
While this keeps the HTTP API from crashing, I'm not sure if this will make it into the next release. There's currently an issue with the iRODS server forcefully killing the connections (i.e. the iRODS server doesn't touch the pooled connections).

This may be a bug in the shutdown logic of the iRODS server.

Reply:
If the iRODS client is able to tell the server it is going away (if it gets the chance to do that)... then the server should terminate the agent servicing the connection.   Are we trying really hard to let the iRODS server know before we die?

Reply:
This is the case where the iRODS server disappears. The iRODS server doesn't kill the agents for some reason which results in the HTTP API being able to still do work.

Reply:
ah, i see now.

Reply:
I was able to confirm that killing a pooled connection is handled gracefully. Below is what I did ...
1. Launched the HTTP API with one connection.
2. Used `kill` to terminate the agent representing the pooled connection.
3. Used `curl` to authenticate. This triggered a SIGPIPE message in the HTTP API, but the connection pool recovered immediately without the client noticing. The token was returned.
4. Used `curl` to do a partial read of a data object and everything worked. No other errors appeared.

So I'm sure the SIGPIPE solution/commit is correct. We just need to figure out what the HTTP API should do when the iRODS server disappears.

Reply:
excellent.   yes, HTTP API should definitely just 'try again' if its connection is bad for any reason... perhaps with a timeout/retrycount before it starts to log things as angry.

Reply:
Well, the real answer is for the iRODS server to kill the agents. The HTTP API shouldn't need to do anything.

Reply:
Right, the Agents should be reaped.  Then the HTTP API would just reconnect as needed.

Reply:
This is fine to include.